src/mlpack/methods/ne/species.hpp

```diff
@@ -0,0 +1,141 @@
+/**
+ * @file species.hpp
+ * @author Bang Liu
+ *
+ * Definition of Species class.
+ */
+#ifndef MLPACK_METHODS_NE_SPECIES_HPP
+#define MLPACK_METHODS_NE_SPECIES_HPP
+
+#include <cstddef>
+
+#include <mlpack/core.hpp>
+
+#include "link_gene.hpp"
+#include "neuron_gene.hpp"
+#include "genome.hpp"
+
+namespace mlpack {
+namespace ne {
+
+/**
+ * This class defines a species of genomes.
+ */
+class Species {
+ public:
+  // Genomes.
+  std::vector<Genome> aGenomes;
+
+  // Default constructor.
+  Species() {
+    aId = -1;
+    aStaleAge = -1;
+    aBestFitness = DBL_MAX;  // DBL_MAX denotes haven't evaluate yet.
+    aNextGenomeId = 0;
+  }
+
+  // Parametric constructor.
+  Species(Genome& seedGenome, int speciesSize) {
+    aId = 0;
+    aStaleAge = 0;
+    aBestFitness = DBL_MAX; 
+    aNextGenomeId = speciesSize;
+
+    // Create genomes from seed Genome and randomize weight.
+    for (int i=0; i<speciesSize; ++i) {
+      Genome genome = seedGenome;
+      genome.Id(i);
+      aGenomes.push_back(genome);
+      aGenomes[i].RandomizeWeights(-1, 1);
+    }
+  }
+
+  // Destructor.
+  ~Species() {}
+
+  // Operator =.
+  Species& operator =(const Species& species) {
+    if (this != &species) {
+      aId = species.aId;
+      aStaleAge = species.aStaleAge;
+      aBestFitness = species.aBestFitness;
+      aBestGenome = species.aBestGenome;
+      aNextGenomeId = species.aNextGenomeId;
+      aGenomes = species.aGenomes;
+    }
+
+    return *this;
+  }
+
+  // Set id.
+  void Id(int id) { aId = id; }
+
+  // Get id.
+  int Id() const { return aId; }
+
+  // Set age.
+  void StaleAge(int staleAge) { aStaleAge = staleAge; }
+
+  // Get age.
+  int StaleAge() const { return aStaleAge; }
+
+  // Set best fitness.
+  void BestFitness(double bestFitness) { aBestFitness = bestFitness; }
+
+  // Get best fitness.
+  double BestFitness() const { return aBestFitness; }
+
+  // Get species size.
+  int SpeciesSize() const { return aGenomes.size(); }
+
+  // Set best fitness to be the minimum of all genomes' fitness.
+  void SetBestFitnessAndGenome() {
+    if (aGenomes.size() == 0) 
+      return;
+
+    aBestFitness = aGenomes[0].Fitness();
+    for (int i=0; i<aGenomes.size(); ++i) {
+      if (aGenomes[i].Fitness() < aBestFitness) {
+        aBestFitness = aGenomes[i].Fitness();
+        aBestGenome = aGenomes[i];
+      }
+    }
+  }
+
+  // Sort genomes by fitness. First is best.
+  static bool CompareGenome(Genome lg, Genome rg) {
```

Comment:
We should use a reference here, to avoid a copy, also it might be useful to use const here.